src/code/z_room.c

```diff
@@ -52,170 +52,207 @@ void Room_Draw0(PlayState* play, Room* room, u32 flags) {
         gSPMatrix(POLY_XLU_DISP++, &gMtxClear, G_MTX_MODELVIEW | G_MTX_LOAD);
     }
 
-    meshHeader0 = &room->meshHeader->meshHeader0;
-    meshHeader0Entry = SEGMENTED_TO_VIRTUAL(meshHeader0->entries);
-    for (i = 0; i < meshHeader0->numEntries; i++) {
-        if ((flags & ROOM_DRAW_OPA) && (meshHeader0Entry->opa != NULL)) {
-            gSPDisplayList(POLY_OPA_DISP++, meshHeader0Entry->opa);
+    meshHeaderAll = &room->meshHeader->all;
+    dListsEntry = SEGMENTED_TO_VIRTUAL(meshHeaderAll->entries);
+    for (i = 0; i < meshHeaderAll->numEntries; i++) {
+        if ((flags & ROOM_DRAW_OPA) && (dListsEntry->opa != NULL)) {
+            gSPDisplayList(POLY_OPA_DISP++, dListsEntry->opa);
         }
 
-        if ((flags & ROOM_DRAW_XLU) && (meshHeader0Entry->xlu != NULL)) {
-            gSPDisplayList(POLY_XLU_DISP++, meshHeader0Entry->xlu);
+        if ((flags & ROOM_DRAW_XLU) && (dListsEntry->xlu != NULL)) {
+            gSPDisplayList(POLY_XLU_DISP++, dListsEntry->xlu);
         }
 
-        meshHeader0Entry++;
+        dListsEntry++;
     }
 
     CLOSE_DISPS(play->state.gfxCtx, "../z_room.c", 239);
 }
 
-#define SHAPE_SORT_MAX 64
-
-typedef struct struct_80095D04 {
-    /* 0x00 */ MeshHeader2Entry* unk_00;
-    /* 0x04 */ f32 unk_04;
-    /* 0x08 */ struct struct_80095D04* unk_08;
-    /* 0x0C */ struct struct_80095D04* unk_0C;
-} struct_80095D04; // size = 0x10
-
-void Room_Draw2(PlayState* play, Room* room, u32 flags) {
-    MeshHeader2* meshHeader2;
-    MeshHeader2Entry* meshHeader2Entry;
-    struct_80095D04 spB8[SHAPE_SORT_MAX];
-    struct_80095D04* spB4 = NULL;
-    struct_80095D04* spB0 = NULL;
-    struct_80095D04* iter;
+typedef struct MeshHeaderCullEntryLinked {
+    /* 0x00 */ MeshHeaderCullEntry* entry;
+    /* 0x04 */ f32 startZ;
+    /* 0x08 */ struct MeshHeaderCullEntryLinked* prev;
+    /* 0x0C */ struct MeshHeaderCullEntryLinked* next;
+} MeshHeaderCullEntryLinked; // size = 0x10
+
+/**
+ * Handle room drawing for the "cull" type of mesh header.
+ *
+ * Each entry referenced by the header is attached to display lists, and a position and radius indicating the volume
+ * those display lists take.
+ * The first step is Z-sorting the entries, also excluding the entries which bounding sphere is entirely before or
+ * beyond the rendered depth range.
+ * The second step is drawing the entries that are left, in ascending depth.
```

Comment:
Nice, this is much better now. Two things
```suggestion
 * The first step Z-sorts the entries, and excludes those entries with a bounding sphere that is entirely before or
 * beyond the rendered depth range.
 * The second step is drawing the entries that are left, from farthest to nearest.
```
(could also use "without" or "outside" the rendered depth range).

"ascending depth" is a bit oxymoronic, I think this is a bit clearer although perhaps it's a little far from the z-depth idea.

Reply:
I did the inanimate-whose "which" again
not sure where I learned this mistake from :thinking: 

Also it's not `from farthest to nearest`, it's the opposite, it's drawing from the nearest entry
Performance-wise, this is probably so that the rdp can ditch triangles that are obscured by geometry just by checking the depth framebuffer instead of overwriting the same pixel in the color framebuffer over and over

Reply:
Ah right, that makes more sense. I'm more used to thinking about it like painting, where you do the sky first, then the mountains, then the happy little trees and so on...

Reply:
That said, it's good to draw opa dlists from nearest to furthest probably, but it would be better to draw the xlu dlists from furthest to nearest to have things look better. OoT doesn't use enough partial translucency that this really matters though